bolt-modules/boltlib/lib/puppet/functions/download_file.rb

```diff
@@ -0,0 +1,104 @@
+  #   download_file('/etc/ssh', 'ssh', $targets)
+  # @example Download a file from multiple Linux targets and compare its contents to a local file
+  #   $results = download_file($source, $destination, $targets)
+  #
+  #   $local_content = file::read($source)
+  #
+  #   $mismatched_files = $results.filter |$result| {
+  #     $remote_content = file::read($result['path'])
+  #     $remote_content == $local_content
+  #   }
+  dispatch :download_file do
+    scope_param
+    param 'String[1]', :source
+    param 'String[1]', :destination
+    param 'Boltlib::TargetSpec', :targets
+    optional_param 'Hash[String[1], Any]', :options
+    return_type 'ResultSet'
+  end
+
+  # Download a file or directory, logging the provided description.
+  # @param source The absolute path to the file or directory on the target(s).
+  # @param destination The path to the destination directory on the local system. Expands relative
+  #                    to `<project>/downloads/`.
+  # @param targets A pattern identifying zero or more targets. See {get_targets} for accepted patterns.
+  # @param description A description to be output when calling this function.
+  # @param options A hash of additional options.
+  # @option options [Boolean] _catch_errors Whether to catch raised errors.
+  # @option options [String] _run_as User to run as using privilege escalation.
+  # @return A list of results, one entry per target, with the path to the downloaded file under the
+  #         `path` key.
+  # @example Download a file from multiple Linux targets to a destination directory
+  #   download_file('/etc/ssh/ssh_config', '~/Downloads', $targets, 'Downloading remote SSH config')
+  dispatch :download_file_with_description do
+    scope_param
+    param 'String[1]', :source
+    param 'String[1]', :destination
+    param 'Boltlib::TargetSpec', :targets
+    param 'String', :description
+    optional_param 'Hash[String[1], Any]', :options
+    return_type 'ResultSet'
+  end
+
+  def download_file(scope, source, destination, targets, options = {})
+    download_file_with_description(scope, source, destination, targets, nil, options)
+  end
+
+  def download_file_with_description(_scope, source, destination, targets, description = nil, options = {})
+    unless Puppet[:tasks]
+      raise Puppet::ParseErrorWithIssue
+        .from_issue_and_stack(Bolt::PAL::Issues::PLAN_OPERATION_NOT_SUPPORTED_WHEN_COMPILING, action: 'download_file')
+    end
+
+    options = options.select { |opt| opt.start_with?('_') }.transform_keys { |k| k.sub(/^_/, '').to_sym }
+    options[:description] = description if description
+
+    executor = Puppet.lookup(:bolt_executor)
+    inventory = Puppet.lookup(:bolt_inventory)
+
+    # Paths should expand relative to the default downloads directory for the project
+    # e.g. ~/.puppetlabs/bolt/downloads/
+    project_downloads = Puppet.lookup(:bolt_project_data)&.downloads
+    destination = File.expand_path(destination, project_downloads)
```

Comment:
We should validate that the destination doesn't have any directory traversal going on. It may be better to restrict it to a specific set of characters (like a download "key" rather than a path).

I'm on the fence about whether we should allow it to be nested. On the one hand, that seems like it opens up even more issues around directory traversal. But it might also be helpful if you have multiple downloads in a plan (`myplan/download1`, `myplan/download2`)... Maybe for now it's best to not allow that and we can open it up later.

Reply:
Just to make sure I understand, by no directory traversal you mean the destination is not something nested more than 1 level deep inside `downloads`? I.e. `foo` as a destination is valid, but `foo/bar` is not?

If that's the case, do we not want to permit absolute paths for a destination and force plans to always download to `<project>/downloads`?

Reply:
I was mostly thinking that it shouldn't contain characters like `..` to be able to "escape" the `downloads` directory. I'm not sure about the nested case.

I don't think absolute paths make a whole lot of sense though, given that we don't exactly download files to the specified directory (but instead to our own structure). It might be better to hold off on absolute paths for now, we can always add them later if needed.

Reply:
Ah, that makes sense. :) I'll add a check for that.

Reply:
Updated.